File: revLL.java
// Time Complexity :O(n) n-number of nodes
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :-
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode rev=head;
        ListNode temp=null;
        
        while(head!=null)
        {
            rev=head;
            head=head.next;
            rev.next=temp;
            temp=rev;
            
        }
        return rev;
    }
}

//Recursion
/*
class Solution {
    public ListNode reverseList(ListNode head) {
     
      
      return helper(head);
    }
    private ListNode helper(ListNode head)
    {
        
        if(head==null || head.next==null) return head; 
        ListNode p=helper(head.next);
        head.next.next=head;
        head.next=null;
        return p;
    }
}
*/

File: RevLL.py
#TimeComplexity:O(n)
#SpaceComplexity:O(n)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this :No
# Used three pointers to reverse the linked List


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head == None:
            return head
        
        curr = head
        
        prev = None
    
        
        fast = curr.next
        
        while(fast != None):
            
            curr.next = prev
            prev = curr
            curr = fast 
            fast = fast.next
        curr.next = prev
        
        return curr